src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/Flows/AcquireTokenHandlerBase.cs

```diff
@@ -332,12 +332,23 @@ protected virtual void PostTokenRequest(AuthenticationResultEx result)
 
         private async Task<AuthenticationResultEx> SendHttpMessageAsync(IRequestParameters requestParameters)
         {
-            client = new AdalHttpClient(this.Authenticator.TokenUri, this.CallState)
+            var metadata = await InstanceDiscovery.GetMetadataEntry(
+                new Uri(Authenticator.TokenUri).Host, Authenticator.ValidateAuthority, this.CallState).ConfigureAwait(false);
+            client = new AdalHttpClient(ReplaceHost(Authenticator.TokenUri, metadata.PreferredNetwork), this.CallState)
```

Comment:
would it not be easier to make this decision in AdalHttpClient class? That way the request classes stay agnostic to details of the metadata and the change applies to all cases?

Reply:
That was also my first reaction when I started considering on this direction. But then something pops up:

* There are 7 files in the "src/.../Http" folder of this repo, and at least 2 files seem relevant, the [AdalHttpClient](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/blob/rayluo/lwn-mso-migration-preferred-network-2/src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/Http/AdalHttpClient.cs) and its underlying helper, the [HttpClientWrapper](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/blob/rayluo/lwn-mso-migration-preferred-network-2/src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/Http/HttpClientWrapper.cs). So one of them could be a candidate to receive this PR change. But I/we can not easily decide which one, if at all, is the suitable place to receive changes, without reading through the entire 33000+ line code base to know what other behaviors/components in this code base are also using those HttpClient classes.
* In fact I already know the InstanceDiscovery is using AdalHttpClient to talk to a given host. If we would change the AdalHttpClient behavior, that could mean the InstanceDiscovery host would also be alterred to the PreferredNetwork. A subtle change of behavior, and probably not what we want.

So I go with the safer way, to add the PreferredNetwork behavior into the exact place where we want it happen, rather than elsewhere.

Reply:
InstanceDiscovery should use HttpClientWrapper directly and everything else will fall in place if you move this logic in AdalHttpClient